Задание 2

MariaOvs opened the pull request:
Задания по второй лабораторной (2.1 и 2.2)

Commit:
Задание 2

Задания по второй лабораторной (2.1 и 2.2)

shishkin-pavel commented on nsu-sdt2020/src/nsu_sdt2020/Task2_1.clj:
```diff
@@ -0,0 +1,48 @@
+(ns Task2_1)
+
+(def ^:const step 0.1)
```
хотелось бы и шаг передавать аргументом

shishkin-pavel commented on nsu-sdt2020/src/nsu_sdt2020/Task2_2.clj:
```diff
@@ -0,0 +1,53 @@
+(defn -main
+  [& args]
+
+  (let [x 1]      (time(print-result 1 ((integral (fn [x] (* x 2))) x))))
+  (let [x 1]      (time(print-result 1 ((integral (fn [x] (* x 2))) x)))) ; =>  прирост производительности
+  (let [x 3]      (print-result 9 ((integral (fn [x] (* x 2))) x)))
+  (let [x 10]     (print-result 100 ((integral (fn [x] (* x 2))) x)))
+
+  (let [x 1]      (print-result (/ 1.0 3) ((integral (fn [x] (* x x))) x)))
+  (let [x 3]      (print-result 9 ((integral (fn [x] (* x x))) x)))
+  (let [x 10]     (print-result (/ 1000.0 3) ((integral (fn [x] (* x x))) x)))
+
+  (let [x 1]      (print-result 5.5 ((integral (fn [x] (+ x 5))) x)))
+  (let [x 3]      (print-result 19.5 ((integral (fn [x] (+ x 5))) x)))
+  (let [x 10]     (print-result 100 ((integral (fn [x] (+ x 5))) x)))
```
не проверял, но ставлю на то, что прирост производительности тут обусловлен прогревом jvm, а не использованием ленивых последовательностей
хочется получить объяснение, почему так :)

shishkin-pavel replied:
проверять можно на значениях наподобии 0...1000, 0...999 (должно отработать почти моментально), 0...1001 (должно работать мало), 0...2000 (примерно сравнимо с первым)

shishkin-pavel commented on nsu-sdt2020/src/nsu_sdt2020/Task2_1.clj:
```diff
@@ -0,0 +1,48 @@
+(defn count-trapeze [f start end]
+  (*    (/ (+ (f start) (f end)) 2)      (- end start)))
+
+(def integrate
+  (memoize
```
почему это плохо?

shishkin-pavel reviewed (changes requested):
в 2.2 нужно что-то поменять
для сдачи нужно коротко переговорить про эти лабы

Commit:
Измененное задание 2

Внесла изменения (по проблемам в пул-реквесте) отностиьельно мемоизации, шага в качестве параметра,первого вызова при прогреве JVM

shishkin-pavel commented on nsu-sdt2020/src/nsu_sdt2020/Task2_1.clj:
```diff
@@ -17,6 +18,23 @@
+(def ^:const eps 0.1)
+
+(def count-trapeze
+  (memoize (  fn [f start end] (*    (/ (+ (f start) (f end)) 2)      (- end start)))))
```
теперь мемоизируются только площади отдельных трапеций, а для получения результата необходимо сделать много сложений (площадей этих трапеций)

shishkin-pavel commented on nsu-sdt2020/src/nsu_sdt2020/Task2_2.clj:
```diff
@@ -49,5 +68,31 @@
   (let [x 3]      (print-result 19.5 ((integral (fn [x] (+ x 5))) x)))
   (let [x 10]     (print-result 100 ((integral (fn [x] (+ x 5))) x)))
+=======
+  (let [x 1]      (time(print-result 1 ((integral (fn [x] (* x 2)) 0.01) x)))) ;происходит прогрев JVM
```
и всё-таки мне кажется, что сейчас между разными вызовами интегрирования прироста производительности (в результате вычисления ленивой последовательности) не должно быть видно

Commit:
Update Task2_2.clj

Исправлена работа с ленивыми последовательностями
